### PR TITLE
DRISE open-source ReImplementation

### PR DESCRIPTION
**Pull request by vbhavank**
The following PR is responsible for implementing an open-source version of [D-RISE algorithm](https://arxiv.org/abs/2006.03204) for explaining black-box detectors using saliency maps.

### CODE REVIEW

File: xaitk_saliency/impls/gen_detection_sal/DRISE.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+        if len(perturb_masks) != len(perturbed_dets):
+            raise ValueError("Number of perturbation masks and respective",
+                 "detections vector do not match.")
+
+        self.n_targets = len(ref_dets)
+        n_targets = len(ref_dets)
+        self.N = len(perturb_masks)
+        self.input_size = perturb_masks[0].shape
+        self.masks = torch.from_numpy(perturb_masks)
```
Comment (Purg) [inline xaitk_saliency/impls/gen_detection_sal/DRISE.py:96]: Does this need to use torch for these operations? While this is an implementation an may require optional dependencies, it seems that this is just performing array operations that should have numpy equivalents. If so, then we should reduce the torch dependency down to numpy (which is already required).

File: examples/DRISE.ipynb [hunk 2]
```diff
@@ -0,0 +1,523 @@
...
...Twem08jt9Q5jnKg3W/BojFk4gLnS9xviJBlhtfnRnPekEJimGaMFTJqzjBxqycRlJoWMKolh8AyDofcaZxUQLz4Xpa3ErIDs8ADrNMbKbrKq3HiQmVwWQlgwrsMohdGyDytFM84JSsVoIXa4bsOmt1hT8YBXCmpGOU1GsFYaDU31mhFq+DiO6BCIyyzUeW1lca2EF2q7nqEf2PaOHKXLiHEi5wXnZV/XdR1hnITBqCsxL1gMu92GsCws00xcIrXI1yxaDmhhWGZUlc5dOXM58LfbDRjHeZrY9gNhHNlsB7TR/M3f/A2lqMvnLS6RmCL73Ybz+cybt28pJXN8OnDz8iUf7x74/Z++4e7hgLNbfv9Pf5BIexTRa54OJ/quI6SA9YqilJD9k2DUtkOPdpZu6HGdB60YxwlQ5GURn6bikijQ9QPXqiWWq9rUhlIsP/sooWqFsU7EOEVR0RL6qD1LrCSgGzb0LeizNFxbqoXS0qWDKsSwsEx/5PD0yA/vf+C//P1/xRoJW+2c/K2ZKTlSi4wJ9SpJ16LQE9apmN9z+xzLBSAouhQTzgler9YsXqwksPAUBXxgrMNqSVM3iIVCG9nfhmWh871E++SMMQMFKZpX8LcwaaXAX0UZcl60aVTDN+UiHFYJYfVy+dbSxEXy+1RLZPbey/Oe8kWIUj45iy/dxI91cV3v3+I3L6h6QHU9xVkyYrZVpbn3D++o9w/Y4QZ/dQvdIL4TKjUuUAppCbhuQHcDVSnKfELrSp5FiumGvdzGaWE5fOTw/hu++cPf88O333J4jCzREXOWoEKjiAlQ4rlZQx4lPkBf5JZaa0ncrOXif8rN+2AkrpeQA/1mkC5odbOjMNWgmkFYGUVsF0Up9eJ9UEohuXXywahJRpDWChqmlopRhhwr2ni860hJ0CpoMTXP89TUjBGUwn3C7FJaVDirwrAUuaTWMVfU4WIB0E2tuYI0oV6AmcYO+K4jNvWQt4ZpiZAlBbYAUSlyrCSKfK/W0fdD88pBqRJJv2ihcy8x8ubNK774vMdZy939I09PJ5ZFRgLzUvH9NR8+zHh34uVLxfX1QN/1PN7DMs0ctcIYSGnmcJ7ofMdQNtQqF1TvNAuRMD4RSmzg0QZFLrILuggzSiGHGUrAGY3ZiIzfeo0yki4NDm2laxAmoiCrtO6piApVK+Ha1RxJqilIq0WpfImN
```
Comment (brianhhu) [inline examples/DRISE.ipynb:283]: D-RISE requires the class probabilities, not just the class label as returned by torchvision detection models. This might require some additional model surgery to get the logits (vs. just the class label).

**Review by brianhhu — COMMENTED** [review]
Adding some notes for the next revision of the code.

File: xaitk_saliency/impls/gen_detector_prop_sal/drise_scoring.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        perturbed_dets: np.ndarray,
+        perturbed_masks: np.ndarray,
+    ) -> np.ndarray:
+
+        if len(perturbed_dets) != len(perturbed_masks):
+            raise ValueError("Number of perturbation masks and respective",
+                             "detections vector do not match.")
+
+        if ref_dets.shape[1] != perturbed_dets.shape[2]:
+            # TODO: This is not necessarily classes, could also be mismatch
```
[the remark this answers is not in the document]
Comment (brianhhu) [reply xaitk_saliency/impls/gen_detector_prop_sal/drise_scoring.py:86]: Essentially, this dimension check (`ref_dets.shape[1]` and `perturbed_dets.shape[2]`) should be `nClasses + 4 + 1`.  However, if there is a shape mismatch it could one of many things: 1) `nClasses` don't agree, 2) one of them is missing the objectness score, or 3) bounding box dimensions are improperly defined. So the `ValueError` of "Number of classes don't match" is not entirely accurate. This was a reminder to change the error message, which I've done so now.

File: xaitk_saliency/impls/gen_detector_prop_sal/drise_scoring.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        sal = weight_regions_by_scalar(s, perturbed_masks)
+
+        # Normalize final saliency map
+        sal = maxabs_scale(
+            sal.reshape(sal.shape[0], -1),
+            axis=1
+        ).reshape(sal.shape)
+
+        # Ensure saliency map in range [-1, 1]
+        sal = np.clip(sal, -1, 1)
```
[the remark this answers is not in the document]
Comment (brianhhu) [reply xaitk_saliency/impls/gen_detector_prop_sal/drise_scoring.py:125]: True, although we do this in the other implementations (I guess to be extra cautious?) So I'll just it keep it here for now for consistency.

File: examples/DRISE.ipynb [hunk 5]
```diff
@@ -0,0 +1,518 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "bbffe8fe",
+   "metadata": {},
+   "source": [
+    "# Black-box Detector\n",
+    "In this example we will use a PyTorch-based pretrained Faster-RCNN object detector with a ResNet50 backbone.\n",
+    "This model normally outputs a dictionary of predicted bounding boxes, scores, and labels. However, since we will be using the [D-RISE](https://arxiv.org/abs/2006.03204) algorithm which requires class probabilities for computing similarity, we'll hack the original implementation to output these by monkey patching the bounding box `postprocess_detections` function."
```
[the remark this answers is not in the document]
Comment (brianhhu) [reply examples/DRISE.ipynb:201]: Yes, this is a good point. I've changed it to be more about the `GenerateDetectorProposalSaliency` interface, and noted that the D-RISE algorithm also makes use of this pattern.

File: tests/impls/gen_detector_prop_sal/test_drise_scoring.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+    def test_get_config(self) -> None:
+        """
+        Test expected configuation behavior.
+        """
+        impl = DRISEScoring('euclidean')
+        for i in configuration_test_helper(impl):
+            assert i.proximity_metric == 'euclidean'
+
+    def test_metric_args(self) -> None:
```
[the remark this answers is not in the document]
Comment (brianhhu) [reply tests/impls/gen_detector_prop_sal/test_drise_scoring.py:36]: Done

File: tests/impls/gen_detector_prop_sal/test_drise_scoring.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        """
+        impl = DRISEScoring('hamming')
+        assert impl.proximity_metric == 'hamming'
+
+    def test_shape_sanity(self) -> None:
+        """
+        Test basic scoring with a single feature for broadcasting sanity check.
+        """
+        impl = DRISEScoring()
+        np.random.seed(2)
```
[the remark this answers is not in the document]
Comment (brianhhu) [reply tests/impls/gen_detector_prop_sal/test_drise_scoring.py:60]: I took a quick look at the link, but was unclear about how to proceed. Do I just initialize an instance of `DRISEScoring` with the random seed set (e.g. `self.impl = DRISEScoring()` and `np.random.seed(2)`), and then use that internal instance everywhere?

File: tests/impls/gen_detector_prop_sal/test_drise_scoring.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        impl = DRISEScoring('hamming')
+        assert impl.proximity_metric == 'hamming'
+
+    def test_bad_metric_args(self) -> None:
+        """
+        Test non-default metric type.
+        """
+        with pytest.raises(
+            ValueError,
+            match=r"Chosen comparision metric not supported or "
```
Comment (Purg) [inline tests/impls/gen_detector_prop_sal/test_drise_scoring.py:50]:
```suggestion
            match=r"Chosen comparison metric not supported or "
```

File: xaitk_saliency/impls/gen_detector_prop_sal/drise_scoring.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+    https://arxiv.org/abs/2006.03204
+    """
+
+    def __init__(
+        self,
+        proximity_metric: str = 'cosine'
+    ):
+
+        try:
+            # Attempting to use chosen comparision metric
```
Comment (Purg) [inline xaitk_saliency/impls/gen_detector_prop_sal/drise_scoring.py:34]:
```suggestion
            # Attempting to use chosen comparison metric
```

File: xaitk_saliency/impls/gen_detector_prop_sal/drise_scoring.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+            cdist([[1], [1]], [[1], [1]], proximity_metric)
+            self.proximity_metric: str = proximity_metric
+        except ValueError:
+            raise ValueError("Chosen comparision metric not supported or "
```
Comment (Purg) [inline xaitk_saliency/impls/gen_detector_prop_sal/drise_scoring.py:38]:
```suggestion
            raise ValueError("Chosen comparison metric not supported or "
```

File: tests/impls/gen_detector_prop_sal/test_drise_scoring.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Purg) [reply tests/impls/gen_detector_prop_sal/test_drise_scoring.py:60]: Just the `np.random.seed(2)` part because it is more setting up a per-test background state. Different tests use different initialiations of DRISEScoring, so just retaining one common construction is not really applicable here.

[the remark this answers is not in the document]
Comment (brianhhu) [reply tests/impls/gen_detector_prop_sal/test_drise_scoring.py:60]: Can I propose removing the `np.random.seed(2)` call? I think we can take it out without any loss of generality, since setting the seed isn't critical for testing the results (they are essentially shape and dimension checks where the entries of the matrices themselves are not critical).

[the remark this answers is not in the document]
Comment (Purg) [reply tests/impls/gen_detector_prop_sal/test_drise_scoring.py:60]: Sure, that's probably OK for now. If there are any hickups in testing/behavior we can revisit this.

File: tests/impls/gen_detector_prop_sal/test_drise_scoring.py (hunk 8 above)
Comment (brianhhu) [reply tests/impls/gen_detector_prop_sal/test_drise_scoring.py:50]: Thanks for catching this typo! I've fixed this and the other instances below.

File: tests/impls/gen_detector_prop_sal/test_drise_scoring.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (brianhhu) [reply tests/impls/gen_detector_prop_sal/test_drise_scoring.py:60]: I ran the test locally, and nothing came up- thanks!

**Review by Purg — APPROVED** [review]
LGTM
